add clair3 and whashap

YiJin-Xiong opened the pull request:
<!--
# nf-core/methylong pull request

Many thanks for contributing to nf-core/methylong!

Please fill in the appropriate checklist below (delete whatever is not relevant).
These are the most common things requested on pull requests (PRs).

Remember that PRs should be made against the dev branch, unless you're preparing a pipeline release.

Learn more about contributing: [CONTRIBUTING.md](https://github.com/nf-core/methylong/tree/master/.github/CONTRIBUTING.md)
-->

## PR checklist

- [ ] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
- [x] If you've added a new tool - have you followed the pipeline conventions in the [contribution docs](https://github.com/nf-core/methylong/tree/master/.github/CONTRIBUTING.md)
- [ ] If necessary, also make a PR on the nf-core/methylong _branch_ on the [nf-core/test-datasets](https://github.com/nf-core/test-datasets) repository.
- [x] Make sure your code lints (`nf-core pipelines lint`).
- [x] Ensure the test suite passes (`nextflow run . -profile test,docker --outdir <OUTDIR>`).
- [x] Check for unexpected warnings in debug mode (`nextflow run . -profile debug,test,docker --outdir <OUTDIR>`).
- [ ] Usage Documentation in `docs/usage.md` is updated.
- [x] Output Documentation in `docs/output.md` is updated.
- [x] `CHANGELOG.md` is updated.
- [ ] `README.md` is updated (including new tool citations and authors/contributors).

YiJin-Xiong commented on modules/local/whatshap/haplotag/main.nf:
```diff
@@ -0,0 +1,58 @@
+process WHATSHAP_HAPLOTAG {
```
I'd be happy to contribute it. :)

YiJin-Xiong commented on subworkflows/local/shared_snvcall_clair3/main.nf:
```diff
@@ -0,0 +1,68 @@
+        .join(SAMTOOLS_FAIDX.out.fai)
+        .map { meta, bam, bai, _ref, _fai ->
+        def packaged_model = meta.method ==  "ont" ? "r1041_e82_400bps_sup_v500" : "hifi_revio"
+        def platform = meta.method ==  "ont" ? "ont" : "hifi"
```
When `meta.method == "ont"`, `platform="ont"`, otherwise `"hifi"` ，will `platform` be correctly assigned as `"ont"` after modifying?

nschan replied:
You are correct, sorry for causing confusion, in this case this will return true.

YiJin-Xiong replied:
Yes, u are right! This would be more appropriate.

YiJin-Xiong commented on subworkflows/local/shared_snvcall_clair3/main.nf:
```diff
@@ -0,0 +1,68 @@
+    input
+        .join(SAMTOOLS_FAIDX.out.fai)
+        .join(CLAIR3.out.vcf)
+        .map { meta, bam, bai, ref, fai, vcf -> [meta, bam, bai, ref, fai, vcf] }
```
At first I want to bind the `vcf` output by `clair3` with the `input`, but now I found it's unnecessary, I'll delete this part.

YiJin-Xiong replied:
I messed up what you meant earlier, but now I get it. Yes, you're right, I'll delete `.map{}`.

nschan commented on subworkflows/local/Pacbio_main.nf:
```diff
@@ -133,14 +120,6 @@ workflow PACBIO {
-        pacbio_versions = pacbio_versions.mix(WHATSHAP.out.versions)
     }
     else {
 
```
```suggestion
    else {
       PACBIO_ALIGN_PBMM2(input)
       ch_pile_in = PACBIO_ALIGN_PBMM2.out.ch_pile_in
       pacbio_versions = pacbio_versions.mix(PACBIO_ALIGN_PBMM2.out.versions)
       map_stat = PACBIO_ALIGN_PBMM2.out.flagstat_out

       PACBIO_SPLIT_STRAND_PBCPG_PILEUP(ch_pile_in)
       pacbio_versions = pacbio_versions.mix(PACBIO_SPLIT_STRAND_PBCPG_PILEUP.out.versions)
       if (params.pedgraph) {
            BED2BEDGRAPH(PACBIO_SPLIT_STRAND_PBCPG_PILEUP.out.pile_out)
            pacbio_versions = pacbio_versions.mix(BED2BEDGRAPH.out.versions)
        }
    }
```
I think you probably did not write this, but I dont see a need for this duplication..?

YiJin-Xiong replied:
I didn't write this, but if you want me to optimize it, I can do it.

jkh00 replied:
script looking cleaner now, thanks for the improvement! 😁

nschan commented on subworkflows/local/Pacbio_main.nf:
```diff
@@ -9,9 +9,7 @@ include { PACBIO_ALIGN_PBMM2               } from './pacbio_align_pbmm2/main'
-include { SNVCALL_CLAIR3                   } from './shared_snvcall_clair3/main'
-include { GUNZIP_AWK                       } from './shared_gunzip_awk/main'
-include { WHATSHAP                         } from './shared_whatshap/main'
+include { DOWNSTREAM                       } from './downstream_main'
```
Is `DOWNSTREAM` really descriptive of what is going on in this workflow? The name seems a bit generic to me, but I am not super familiar with the methods.

nschan commented on subworkflows/local/shared_snvcall_clair3/main.nf:
```diff
@@ -0,0 +1,69 @@
+    input
+        .join(SAMTOOLS_FAIDX.out.fai)
+        .map { meta, bam, bai, _ref, _fai ->
+        def packaged_model = meta.method ==  "ont" ? "r1041_e82_400bps_sup_v500" : "hifi_revio"
```
I didn't really think about this earlier, but I wonder if it would be make sense to move this logic inside the module?

nschan replied:
Oh okay, sorry i did not realize that, in that case you can just leave it as it is.

nschan commented on modules/nf-core/clair3/main.nf:
```diff
@@ -0,0 +1,78 @@
+            model = "/usr/local/bin/models/$packaged_model"
+        }
+    }
+    if (!packaged_model) {
```
Wouldn't this be clearer long-term / for others to maintain if you used positive tests, for example something like:

```

if (user_model && packaged_model) error 'Two models specified'

model = (!user_model && !packaged_model) ? 
    "" : user_model ?: 
    (workflow.profile.tokenize(',').intersect(['conda', 'mamba']).size() >= 1) ?
         ${CONDA_PREFIX}/bin/models/${packaged_model}" : "/usr/local/bin/models/${packaged_model}"
```

YiJin-Xiong replied:
I didn't write this, this is an existing nf-core module.

nschan reviewed (commented):
I have added some more comments, I have not really looked at the modules before..

jkh00 commented on subworkflows/local/shared_whatshap/main.nf:
```diff
@@ -0,0 +1,68 @@
+
+    // Prepare inputs for whatshap
+
+    input
+        .map { meta, bam, bai, _ref, _fai, _vcf -> [meta, bam, bai] }
+        .set { ch_bam_in }
+
+    input
+        .map { meta, _bam, _bai, ref, fai, _vcf -> [meta, ref, fai] }
+        .set { ch_ref }
+
+    input
+        .map { meta, _bam, _bai, _ref, _fai, vcf -> [meta, vcf] }
+        .set { ch_vcf }
```
I noticed using single mapping like this might be causing cache issue and not sure if this will pair up the reference correctly to the input. To be on the safe side, you could use multimap to channel them:

```suggestion
input
         .multiMap { meta, bam, bai, ref, fai, vcf ->
                             bam_in: [meta, bam, bai]
                             ref_in: [meta, ref, fai]
                             vcf_in: [meta, vcf]
                        }
          .set { ch_input }

```

and then call them using:
> WHATSHAP_PHASE(ch_input.bam_in, ch_input.ref_in, ch_input.vcf_in) for example

YiJin-Xiong replied:
Thanks for your advice :)

jkh00 commented:
looks good to me! I added some comments but are just minor changes :)

YiJin-Xiong commented on subworkflows/local/ONT_main.nf:
```diff
@@ -90,6 +94,10 @@ workflow ONT {
         }
     }
 
+    DOWNSTREAM(ch_pile_in, ont_versions)
```
ok, putting your ideas together, 
first I'll incorporate `DOWNSTREAM` into `methylong.nf` , `DOWNSTREAM` will include `SNVCALL_CLAIR3`, `GUNZIP_AWK`, `WHATSHAP` and `DMR analysis`. 
then I'll add a `--skip_SNVs` parameter for users to choose whether to run these steps. 

is that right?

YiJin-Xiong replied:
The `DOWNSTREAM` workflow consists of snvcall → phase → DMR analysis. `SNVCALL_CLAIR3` is used for snvcall, while `GUNZIP_AWK and WHATSHAP` are for phase. so I didn't incorporate these three processes into a sub-workflow named SNV. 
Also, part of the input for DMR analysis is derived from the output of `WHATSHAP`, so I prefer to put them all in `DOWNSTREAM`.

jkh00 commented:
thanks for addressing all the comments! should also include '--skip_snvs' in the ci test. u can find it in .github/workflows/ci.yml under the parameters options.

YiJin-Xiong commented:
> thanks for addressing all the comments! should also include '--skip_snvs' in the ci test. u can find it in .github/workflows/ci.yml under the parameters options.

I've included `'--skip_snvs'` in `ci.yml`, is there anything else I should pay attention to?
